### ISSUE TITLE
ArrayInterfaceCore

### ISSUE THREAD

**Issue by ParadaCarleton** [JuliaArrays/ArrayInterface.jl#211]
As mentioned [here](https://github.com/rafaqz/DimensionalData.jl/issues/307), `ArrayInterface` has become a huge dependency to take on. It might make sense to separate out some `ArrayInterfaceCore` that can be loaded faster and contains only basic functionality.

**Comment by grahamas** [comment]
Sounds like:

1. `AbstractArrayInterface` [1]
2. and `ArrayInterfaceCore` 

I think there's a more useful divide. It seems to me like there are two kinds of users:

1. Someone purely trying to develop their own Array type (usually the ideal use-case for `AbstractX` packages). However, even for this person the `AbstractArrayInterface` wouldn't be sufficient in many cases. The existing package provides a lot of useful low-level stuff (`device` etc) that a more advanced developer might want to use without the cost of importing the various high-level dependencies of this package. This is the person who probably wants something like `ArrayInterfaceCore`, for some definition of "basic functionality," and so is the subject of this issue.
2. Someone trying to use one (or more) of the Array types defined in this package. This person wants more than would be included in my understanding of "basic functionality" (maybe e.g. the `FieldArray`) but still a lot less than the full package (e.g. does not want `CUDA`, `CuArray` etc.). This person suggest splitting up ArrayInterface into sub-packages, all depending on `ArrayInterfaceCore`, each contributing one set of Array implementations. This person isn't really relevant to this issue, but sometimes this person is me, so I think it'd be cool. But it should be clear/decided that "basic functionality" won't include this person.

Basically, that's all to say I don't think `AbstractArrayInterface` (the package) would be all that useful, and I think "basic functionality" can mean a lot of different things to different people. From my perspective, I'd like it to mean that when I subtype `AbstractArrayInterface`, the vast majority of necessary and optional methods have working default implementations. Someone should probably write down what it means, though.

Finally, I'm curious how this ties into the long-mentioned merge into base julia.

[1]  the package, not the type--- maybe this is why no one suggested this.

**Comment by ParadaCarleton** [comment]
> What's the measure here? Load time?

I think it should be; anything like an ArrayInterfaceCore should take under 0.1s to load, I think. Is there a breakdown of load time by file? What precisely makes ArrayInterface take so long to load?

**Comment by ChrisRackauckas** [comment]
Julia v1.7?

**Comment by ChrisRackauckas** [comment]
I'm confused. Splitting to different packages and making use of Requires would only increase compile and startup times, not decrease it. I'm not sure why that's being proposed as the solution then to decreasing startup time.

**Comment by chriselrod** [comment]
Why would we need Requires.jl?

**Comment by rafaqz** [comment]
Im not sure its worth doing anything about this given the 1.7 performance improvements. 

Speaking to my small part in triggering this discussion, 0.1 seconds is not an issue at all for DimensionalData.jl. It will be swallowed by the improvements to load times of DD itself in 1.7.

**Comment by chriselrod** [comment]
> What parts are being proposed to move out and what is the v1.7 `using` time drop from it?

A library of stubs should load nearly instantly.

**Comment by ChrisRackauckas** [comment]
How much using time would it cut? Can we make it a sub-package?

**Comment by Tokazama** [comment]
Wouldn't `parent_type` give incorrect results for all of the array types in base unless `ArrayInterface.jl` was loaded? At what point do we tell users they can reliably use the stubs?

**Comment by chriselrod** [comment]
> At what point do we tell users they can reliably use the stubs?

When ArrayInterface is loaded.

**Comment by Tokazama** [comment]
Is the idea that we would similarly cover a generic interface for Base, Core, Statistics, etc., in ArrayInterface (similar to what ChainRules does) or are we just making function stubs and keeping things as is here?

In other words, does this proposal change the scope of ArrayInterface at all?

**Comment by chriselrod** [comment]
The stub library would help us remove the `@requires`, by making it easier to convince other libraries to accept a PR adding the definitions there.

From there, we can find out if anything else is particularly slow.

The full library will always be slower to load, by virtue of having many more lines of code.

**Comment by ChrisRackauckas** [comment]
>The full library will always be slower to load, by virtue of having many more lines of code.

Not necessarily true. Adding more code does not always make a meaningful difference to load time. It's not hard to come up with a 100,000 line package that loads in microseconds. That's why the real question isn't how much code, but which code is causing the loading time? This entire conversation is going down a meaningless tangent of splitting things out without even figuring out what the actual problem is.

**Comment by Tokazama** [comment]
I think at least one thing we all agree on is that we need to move away from Requires. I just submitted a PR to OffsetArrays.jl. I might be able to do one to StaticArrays.jl next week sometime.

**Comment by ChrisRackauckas** [comment]
Yes, Requires.jl has to go as much as possible. In some cases, that might be best done by subpackages like how KernelAbstractions.jl does it. https://github.com/JuliaGPU/KernelAbstractions.jl/tree/master/lib/CUDAKernels is a separate package with separate deps in the same repo. We could split like this and try to downstream those pieces as the using time gets trivialized.

But still, I think someone needs to do experiments to see what matters. If you delete all of the `@requires` code, does the using time go to almost zero? I wouldn't be surprised if that's a yes.

**Comment by Tokazama** [comment]
If I just remove the code related to `BandedMatrices` and `BlockBandedMatrices` (the code in the requires section and `BandedBlockBandedMatrixIndex`,`BandedMatrixIndex`), then I get down to 0.904975 seconds.

**Comment by ChrisRackauckas** [comment]
Yeah, BandedMatrices and BlockBandedMatrices seems to be one of the big startup time hits. @dlfivefifty would you accept those functions downstream to remove the Requires.jl usage?

**Comment by chriselrod** [comment]
I imagine you'll get responses similar to here https://github.com/JuliaArrays/OffsetArrays.jl/pull/268 at many other libraries, including CUDA and StaticArrays.
Hopefully you can persuade them, but I have little interest in arguing/persuading/selling ideas, so my inclination is always to focus on making it the easiest sell I can.

Also, on a relatively fresh master, I'm still seeing 0.7s load times from ArrayInterface:
```julia
julia> @time using ArrayInterface
  0.703473 seconds (1.55 M allocations: 86.349 MiB, 2.57% gc time, 86.84% compilation time)

julia> versioninfo()
Julia Version 1.8.0-DEV.754
Commit b8ed1ae9e7 (2021-10-15 22:03 UTC)
```
This is with `--startup=no`, so that I haven't loaded any packages.
To @ChrisRackauckas's point, using SnoopCompile to find out where that 86% compile time is coming from would be a start to addressing the slow load time.

**Comment by Tokazama** [comment]
Digging through SnoopCompile stuff as I have time. It looks like 65 of 100 unique invalidations come from Static.jl, most of which are `StaticInt` related.

**Comment by chriselrod** [comment]
FWIW
```julia
julia> Aqua.test_ambiguities(ArrayInterface)
Test Passed
  Expression: success(pipeline(cmd; stdout = stdout, stderr = stderr))

julia> versioninfo()
Julia Version 1.8.0-DEV.754
Commit b8ed1ae9e7 (2021-10-15 22:03 UTC)
```

**Comment by Tokazama** [comment]
Currently `AbstractArray2` only serves as a way of helping users not define as many methods manually, but if it's a problem we could probably get rid of `AbstractArray2`. I still think it has merit, but it should probably be implemented down the line as an independent and more complete package that reimplements array method using traits defined here (e.g., `cat`, `*`, etc.).

**Comment by Tokazama** [comment]
Looking at getting `StaticInt` into Base and I realized we would need to include `OptionallyStaticUnitRange` and `OptionallyStaticStepRange`. Otherwise, we would have to get rid of the `:` methods that have `StaticInt`.

**Comment by rafaqz** [comment]
After looking at precompilation time of a lot of packages the last few weeks, the `@require` block here does come up a lot. So the cost of it to the apparent TTFX of other packages is another concern. For example, in DynamicGrids.jl it adds half a second on 1.7:  

With the `@require` block:
```julia
julia> @time using DynamicGrids
  5.363114 seconds (16.95 M allocations: 1.817 GiB, 4.89% gc time, 18.05% compilation time)
```
Commented out:
```julia
julia> @time using DynamicGrids
  4.830753 seconds (15.55 M allocations: 1.744 GiB, 4.64% gc time, 9.16% compilation time)
```

I assume this is for OffsetArrays.jl and StaticArrays.jl. This hidden 0.53 seconds is comparable to the total 0.68 seconds for `using ArrayInterface`.

**Comment by Tokazama** [comment]
I'd really like StaticInt to be in base as a first step here. I haven't had time to bring up a PR to base that will likely have a lot of feedback though.

**Comment by chriselrod** [comment]
> I'd really like StaticInt to be in base as a first step here. I haven't had time to bring up a PR to base that will likely have a lot of feedback though.

Before working on such a PR, I recommend talking to people on the compiler team. Keno, for example, disapproved of my use of it.
I've been told that things such as that are among the contributors to the bad latency of some of my packages and LoopVectorization.jl in particular.

Of course, if it isn't too much trouble, the PR may be a good platform for discussion.

**Comment by timholy** [comment]
> I've been told that things such as that are among the contributors to the bad latency of some of my packages and LoopVectorization.jl in particular.

True, but that landscape may change if we can cache output of LoopVectorization (https://github.com/JuliaLang/julia/pull/43990 almost wipes out the LV-induced latency, at least for that fairly trivial case). Of course that only works for specifically-precompiled types, but I think that will cover quite a lot.

**Comment by timholy** [comment]
Are the oddities similar to https://github.com/JuliaLang/julia/issues/35972? That seems to be fixed by my PR.

Still, the point is well taken. It will be lovely to get big chunks of LV into the compiler, when the time is ripe.

**Comment by chriselrod** [comment]
> Are the oddities similar to [JuliaLang/julia#35972](https://github.com/JuliaLang/julia/issues/35972)? That seems to be fixed by my PR.

Yes, they are. I just built your PR and tried.
Without precompile statements:
```julia
julia> @time TriangularSolve.rdiv!(C, A, UpperTriangular(B), Val(false));
 13.178147 seconds (11.79 M allocations: 697.122 MiB, 2.55% gc time, 100.00% compilation time)

julia> @time TriangularSolve.ldiv!(C, LowerTriangular(B), A, Val(false));
  5.058862 seconds (7.98 M allocations: 498.458 MiB, 8.42% gc time, 100.00% compilation time)

julia> @time TriangularSolve.rdiv!(C, A, UpperTriangular(B));
  0.765654 seconds (1.81 M allocations: 92.006 MiB, 19.53% gc time, 99.94% compilation time)

julia> @time TriangularSolve.ldiv!(C, LowerTriangular(B), A);
  0.103297 seconds (310.55 k allocations: 16.453 MiB, 99.90% compilation time)
```
With:
```julia
julia> @time TriangularSolve.rdiv!(C, A, UpperTriangular(B), Val(false));
  3.000927 seconds (583.43 k allocations: 16.928 MiB, 99.99% compilation time)

julia> @time TriangularSolve.ldiv!(C, LowerTriangular(B), A, Val(false));
  2.739339 seconds (510.76 k allocations: 14.833 MiB, 99.99% compilation time)

julia> @time TriangularSolve.rdiv!(C, A, UpperTriangular(B));
  0.508256 seconds (447.25 k allocations: 22.785 MiB, 23.63% gc time, 99.92% compilation time)

julia> @time TriangularSolve.ldiv!(C, LowerTriangular(B), A);
  0.049398 seconds (3.50 k allocations: 171.672 KiB, 99.86% compilation time)
```
So it made a huge difference.
There were no allocations during `@benchmark`, so the problem seems to be fixed as you suggested. Excellent!

> Still, the point is well taken. It will be lovely to get big chunks of LV into the compiler, when the time is ripe.

I'm slowly working on writing a new LV in C++ that takes LLVM as input.
That should help latency of course, as well as solving other issues, like support for colors/complex numbers/duals/etc.
I'm also adding things like loop fusion, dependency handling, and support for much more complicated loop types.
I wouldn't bet on this arriving any time soon, but it's been the focus of most of my free time for the past few months. I've set the JuliaSIMD ecosystem aside to focus on and start making progress on this.

**Comment by Tokazama** [comment]
As Chris pointed out, it's difficult to know what core devs think is "appropriate" use of the type system and generated functions. I think we've avoided an explosion of trait types here by reusing types from `Static.jl`, but not everybody has the same opinion of what should be a trait. When more of the mysteries of "julia/base/compiler/" become public API it will be easier to really hand some of the stuff here off to base.

**Comment by Tokazama** [comment]
Today I started looking at reducing `using` time here again (Julia v1.7)

This is what we currently have.
```julia
julia> @time using ArrayInterface
  0.469390 seconds (805.63 k allocations: 44.151 MiB, 2.35% gc time, 87.20% compilation time)
```

And if you comment out the `@requires`
```julia
@time using ArrayInterface
  0.060452 seconds (127.12 k allocations: 7.568 MiB, 7.59% compilation time)
```

I recently did some work on improving `Static.jl` so I'm fairly certain that about half of that last one is due to it (~0.03-0.04 seconds).

**Comment by ChrisRackauckas** [comment]
https://github.com/SciML/GalacticOptim.jl now is using the multi-package repo setup, so you could use that as an example of how to put this together.

**Comment by Tokazama** [comment]
Is the idea we have ArrayInterfaceCore.jl contain the substance of everything and ArrayInterface.jl just use requires? Then we can gradually move stuff out of requires?

**Comment by ChrisRackauckas** [comment]
yes, that would also give explicit versioning on the requires'd pieces.

**Comment by Tokazama** [comment]
Gotcha. I'll try looking into it in a couple days.

**Comment by Tokazama** [comment]
Early work on a branch for this [here](https://github.com/JuliaArrays/ArrayInterface.jl/tree/multi-package).

**Comment by oschulz** [comment]
A more lightweight version of ArrayInterface would be awsome, I think - it is nowadays a direct or indirect dependency of a large chunk of the ecosystem. I've been tracking down some package load times, and ArrayInterface often turns out to be a driving factor. With a package like FiniteDiff, for example:

```julia
julia> @time_imports using FiniteDiff
      3.2 ms  ┌ Compat
      0.5 ms  ┌ Requires
    538.4 ms  ┌ StaticArrays
      0.2 ms  ┌ IfElse
     33.1 ms    ┌ Static
    722.1 ms  ┌ ArrayInterface
   1512.6 ms  FiniteDiff
```

ArrayInterface makes up the biggest chunk of the load time, StaticArrays almost all of the rest, FiniteDiff itself is tiny:

```julia
julia> @time_imports using ArrayInterface, StaticArrays
      3.1 ms  ┌ Compat
      0.6 ms  ┌ Requires
      0.1 ms  ┌ IfElse
     32.7 ms  ┌ Static
    935.6 ms  ArrayInterface
    547.9 ms  StaticArrays

julia> @time_imports using FiniteDiff
     15.3 ms  FiniteDiff
```

FIniteDiff (just an example here, though it is itself a direct or indirect dependency of many packages) seems to mostly use `ArrayInterface.allowed_getindex` and `ArrayInterface.allowed_setindex!`, which intuitively shouldn't have to come from a heavy dependency.

**Comment by ChrisRackauckas** [comment]
What is `@time_imports`? That seems awesome.

**Comment by oschulz** [comment]
> What is @time_imports? That seems awesome.

It's new in Julia 1.8 (I've been using v1.8.0-beta3 above). It really opened my eyes about the package load times in general in the last days.

**Comment by oschulz** [comment]
> It really opened my eyes about the package load times in general in the last days.

I want to reduce the horrid load time of BAT.jl by reducing it's dependencies and splitting it up along dependency lines, so I started to measure load times of a bunch of diff- and opt- and stats-related packages and was quite surprised how lightweight some packages are (on top of their deps) and how heavy some others (bat/BAT.jl#351).

**Comment by ChrisRackauckas** [comment]
The split is done. The core is nice. The rest needs work, but hey, that was the point of this issue so closed.

```julia
julia> @time_imports using ArrayInterfaceCore
     28.4 ms  ArrayInterfaceCore

julia> @time_imports using ArrayInterface
      2.8 ms  ┌ Compat
      1.0 ms  ┌ IfElse
     45.1 ms  ┌ Static
      2.3 ms  ┌ ArrayInterfaceCore
    398.1 ms  ArrayInterface

julia> @time_imports using ArrayInterfaceBandedMatrices
    120.3 ms    ┌ FillArrays
   1005.0 ms  ┌ ArrayLayouts
    214.9 ms  ┌ BandedMatrices
      2.3 ms  ┌ ArrayInterfaceCore
   1251.9 ms  ArrayInterfaceBandedMatrices

julia> @time_imports using ArrayInterfaceBlockBandedMatrices
    122.4 ms    ┌ FillArrays
   1017.1 ms  ┌ ArrayLayouts
    207.1 ms  ┌ BandedMatrices
    138.1 ms  ┌ MatrixFactorizations
      2.5 ms  ┌ ArrayInterfaceCore
      1.3 ms  ┌ ArrayInterfaceBandedMatrices
    154.6 ms  ┌ BlockArrays
     41.7 ms  ┌ BlockBandedMatrices
   1595.4 ms  ArrayInterfaceBlockBandedMatrices

julia> @time_imports using ArrayInterfaceCUDA
      2.6 ms  ┌ Compat
      1.5 ms  ┌ Requires
     52.3 ms    ┌ ChainRulesCore
     60.7 ms  ┌ AbstractFFTs
     15.7 ms      ┌ Preferences
     17.6 ms    ┌ JLLWrappers
    173.7 ms  ┌ LLVMExtra_jll
      1.2 ms  ┌ Reexport
     23.5 ms    ┌ RandomNumbers
     38.3 ms  ┌ Random123
      4.0 ms      ┌ CEnum
    100.4 ms    ┌ LLVM
      2.0 ms    ┌ Adapt
    799.5 ms  ┌ GPUArrays
      1.3 ms  ┌ ExprTools
      1.1 ms  ┌ IfElse
     34.2 ms    ┌ Static
      2.3 ms    ┌ ArrayInterfaceCore
    381.2 ms  ┌ ArrayInterface
     24.7 ms  ┌ TimerOutputs
      1.7 ms  ┌ CompilerSupportLibraries_jll
      6.2 ms  ┌ BFloat16s
    193.5 ms  ┌ GPUCompiler
   1059.9 ms  ┌ CUDA
   2855.1 ms  ArrayInterfaceCUDA

julia> @time_imports using ArrayInterfaceOffsetArrays
      2.8 ms  ┌ Compat
      1.3 ms  ┌ IfElse
     30.5 ms    ┌ Static
      2.2 ms    ┌ ArrayInterfaceCore
    327.2 ms  ┌ ArrayInterface
      1.6 ms  ┌ Adapt
     35.4 ms  ┌ OffsetArrays
    405.6 ms  ArrayInterfaceOffsetArrays

julia> @time_imports using ArrayInterfaceStaticArrays
      2.5 ms  ┌ Compat
    460.9 ms  ┌ StaticArrays
      1.1 ms  ┌ IfElse
     33.4 ms    ┌ Static
      2.5 ms    ┌ ArrayInterfaceCore
    333.8 ms  ┌ ArrayInterface
      1.8 ms  ┌ Adapt
    837.3 ms  ArrayInterfaceStaticArrays

julia> @time_imports using ArrayInterfaceTracker
    467.2 ms    ┌ StaticArrays
    478.1 ms  ┌ DiffResults
      2.9 ms  ┌ Compat
      1.3 ms  ┌ NaNMath
      1.3 ms  ┌ Requires
     53.7 ms    ┌ ChainRulesCore
     55.9 ms  ┌ ChangesOfVariables
      1.1 ms  ┌ OpenLibm_jll
      1.5 ms  ┌ InverseFunctions
      4.5 ms  ┌ DocStringExtensions
      5.3 ms    ┌ IrrationalConstants
      1.5 ms    ┌ CompilerSupportLibraries_jll
      2.0 ms      ┌ LogExpFunctions
     36.3 ms          ┌ Preferences
     38.6 ms        ┌ JLLWrappers
    187.3 ms      ┌ OpenSpecFun_jll
    203.6 ms    ┌ SpecialFunctions
     11.1 ms      ┌ MacroTools
     12.9 ms    ┌ CommonSubexpressions
      1.4 ms    ┌ DiffRules
    298.4 ms  ┌ ForwardDiff
      1.8 ms  ┌ Adapt
      2.4 ms  ┌ ArrayInterfaceCore
     26.4 ms  ┌ NNlib
    561.0 ms  ┌ Tracker
   1815.3 ms  ArrayInterfaceTracker
```

**Comment by oschulz** [comment]
Amazing!

**Comment by oschulz** [comment]
So now we just need to ask a bunch of packages to switch deps? :-)

**Comment by ChrisRackauckas** [comment]
CompatHelper will tell them ArrayInterface had a breaking update, and yes most packages should then choose to update to Core.

**Comment by ChrisRackauckas** [comment]
One of the biggest remaining issues is that GPUArrays.jl is still heavy:

https://github.com/JuliaGPU/GPUArrays.jl/issues/409

So any package that removes Requires.jl is not going to cut down on using time until there's a better target than GPUArrays.jl. If that goes Core as well, then anything that's not static is almost instant.

**Comment by oschulz** [comment]
>  and yes most packages should then choose to update to Core.

Could reduce the load time of FiniteDiff by over 50% (the remainder will be StaticArrays, which is harder to address).

**Comment by ChrisRackauckas** [comment]
Yes, I'm doing those PRs now. RecursiveArrayTools first, and a bit of shaming on StaticArrays

**Comment by oschulz** [comment]
> and a bit of shaming on StaticArrays

Amen to that! I did created an issue regarding it's load time (JuliaArrays/StaticArrays.jl#1023) - but to be honest I have to admit that I don't have a concrete proposal on how a lightweight StaticArraysBase.jl or so could look like, without behavior changes when the full one is loaded.

**Comment by ChrisRackauckas** [comment]
See my comment. We only need 7 functions out of the whole thing, and then we delete the second biggest contributor to load times after GPUArrays.jl. Just look at this:

```julia
julia> @time_imports using RecursiveArrayTools
    10.1 ms    ┌ MacroTools
    18.7 ms  ┌ ZygoteRules
    2.6 ms  ┌ Compat
123.9 ms  ┌ FillArrays
521.7 ms  ┌ StaticArrays
    18.4 ms      ┌ Preferences
    20.2 ms    ┌ JLLWrappers
173.5 ms  ┌ LLVMExtra_jll
    4.8 ms      ┌ CEnum
103.7 ms    ┌ LLVM
    2.0 ms    ┌ Adapt
815.3 ms  ┌ GPUArrays
    5.2 ms  ┌ DocStringExtensions
    40.3 ms  ┌ RecipesBase
    2.6 ms  ┌ ArrayInterfaceCore
    56.6 ms  ┌ ChainRulesCore
1822.5 ms  RecursiveArrayTools
```

Fix those two and we go from `1822.5 ms` to `208.3 ms` (since GPUArrays is what also pulls in LLVM). "StaticArrays and GPUArrays not having a core is what causes all of SciML to have an order of magnitude extra startup time" is not an exagguration, it's a proven fact with our tools and data. If those maintainers don't help out out here, than that would be depressing. I hope that we get a quick fix to this now that the cause is so obvious now!

**Comment by oschulz** [comment]
> Fix those two and we go from 1822.5 ms to 208.3 ms

Wow! We should send a big box of whatever is his favourite to @IanButterworth for JuliaLang/julia#41612 !

**Comment by IanButterworth** [comment]
Cool! Nice to see it helping.

If you haven't already, try it in nightly. It reports compilation time and recompilation if any (due to upstream invalidations)

**Comment by oschulz** [comment]
> It reports compilation time and recompilation if any (due to upstream invalidations)

Oh, that could really ... encourage ... package maintainers to take a closer look at a few things. :-)

**Comment by oschulz** [comment]
>  try it in nightly [...] It reports compilation time and recompilation 

Will that still go into 1.8 as well?

**Comment by IanButterworth** [comment]
It was added recently and is not marked for backport, but perhaps that's reasonable.. might be worth suggesting it
